### PR TITLE
[RLlib]: Move OPE to evaluation config

### PR DESCRIPTION
**Pull request by Rohan138**
This PR moves the OPE methods to the evaluation config (and thus the evaluation workers).
Note that I had to remove the OPE from some of the CQL and MARWIL tests because they use evaluation input = "sampler", so using OPE is redundant. However, user code using `config.input_evaluation` should still work with a deprecation warning.

Previous PR: #25899 (already merged)
Next PR: #26279 

## Checks

- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [X] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [X] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/evaluation/metrics.py [hunk 1]
```diff
@@ -147,8 +147,8 @@ def summarize_episodes(
     if new_episodes is None:
         new_episodes = episodes
 
-    episodes, estimates = _partition(episodes)
-    new_episodes, _ = _partition(new_episodes)
+    episodes, _ = _partition(episodes)
```
Comment (sven1977) [inline rllib/evaluation/metrics.py:150]: Was this a bug? Can we add a one-line comment on why we are ignoring the estimates on `episodes`, but not `new_episodes`.

File: rllib/evaluation/rollout_worker.py [hunk 2]
```diff
@@ -735,10 +729,14 @@ def wrap(env):
                     error=False,
                 )
                 method_type = ope_types[method_type]
-            if name == "simulation":
-                logger.warning(
-                    "Requested 'simulation' input evaluation method: "
-                    "will discard all sampler outputs and keep only metrics."
+            if method_type == "simulation":
```
Comment (sven1977) [inline rllib/evaluation/rollout_worker.py:732]: Nice!

File: rllib/evaluation/rollout_worker.py [hunk 3]
```diff
@@ -761,7 +759,7 @@ def wrap(env):
             else:
                 raise ValueError(
                     f"Unknown off_policy_estimation type: {method_type}! Must be "
-                    "either `simulation|is|wis|dm|dr` or a sub-class of ray.rllib."
+                    "either a class path or a sub-class of ray.rllib."
```
Comment (sven1977) [inline rllib/evaluation/rollout_worker.py:762]: Nice. Better to force explicitness and not allow too many shortcut options for the user.

File: rllib/evaluation/rollout_worker.py [hunk 4]
```diff
@@ -785,12 +783,12 @@ def wrap(env):
                 multiple_episodes_in_batch=pack,
                 normalize_actions=normalize_actions,
                 clip_actions=clip_actions,
-                blackhole_outputs="simulation" in off_policy_estimation_methods,
                 soft_horizon=soft_horizon,
                 no_done_at_end=no_done_at_end,
                 observation_fn=observation_fn,
                 sample_collector_class=policy_config.get("sample_collector"),
                 render=render,
+                blackhole_outputs="simulation" in off_policy_estimation_methods,
```
Comment (sven1977) [inline rllib/evaluation/rollout_worker.py:791]: Add TODO to deprecate this once we completely disallow "simulation".

File: rllib/offline/estimators/direct_method.py [hunk 5]
```diff
@@ -16,44 +17,58 @@
...
-    batch is less than `k` or `should_train` is set to False, yields an empty
-    list for train_episodes and all the episodes in test_episodes.
 
+@ExperimentalAPI
+def train_test_split(
+    batch: SampleBatchType,
+    train_test_split_val: float = 0.0,
+    k: int = 0,
+) -> Generator[Tuple[List[SampleBatch]], None, None]:
+    """Utility function that returns either a train/test split or
```
Comment (sven1977) [inline rllib/offline/estimators/direct_method.py:29]: Nit: Our docstrings should always just have a single sentence one-line at the top, then an empty line, then any other information. E.g.:

```
def xyz(..):
    """Some single sentence description.

    Anything else blabla

    Args:
        ...
    """
```

File: rllib/offline/estimators/direct_method.py [hunk 6]
```diff
@@ -16,44 +17,58 @@
...
-            test_episodes = episodes[i * n_fold :]
-        yield train_episodes, test_episodes
+            # Append remaining episodes onto the last eval_episodes
+            eval_episodes = episodes[i * n_fold :]
+        yield eval_episodes, train_episodes
     return
 
 
-@DeveloperAPI
+@ExperimentalAPI
```
Comment (sven1977) [inline rllib/offline/estimators/direct_method.py:71]: Actually, sorry, but let's not use ExperimentalAPI anymore. We are urged by other libraries teams to move RLlib to the generic Ray API annotations (which doesn't have `ExperimentalAPI`). Let's leave as DeveloperAPI.

File: rllib/offline/estimators/fqe_torch_model.py [hunk 7]
```diff
@@ -5,14 +5,14 @@
...
-from ray.rllib.utils.annotations import DeveloperAPI
+from ray.rllib.utils.annotations import ExperimentalAPI
 from ray.rllib.utils.framework import try_import_torch
 from ray.rllib.utils.typing import ModelConfigDict, TensorType
 
 torch, nn = try_import_torch()
 
 
-@DeveloperAPI
+@ExperimentalAPI
```
Comment (sven1977) [inline rllib/offline/estimators/fqe_torch_model.py:15]: same

File: rllib/offline/estimators/importance_sampling.py [hunk 8]
```diff
@@ -2,13 +2,13 @@
...
 )
-from ray.rllib.utils.annotations import override, DeveloperAPI
+from ray.rllib.utils.annotations import override, ExperimentalAPI
 from ray.rllib.utils.typing import SampleBatchType
 from typing import List
 import numpy as np
 
 
-@DeveloperAPI
+@ExperimentalAPI
```
Comment (sven1977) [inline rllib/offline/estimators/importance_sampling.py:11]: same

File: rllib/offline/estimators/off_policy_estimator.py [hunk 9]
```diff
@@ -11,16 +11,16 @@
...
 logger = logging.getLogger(__name__)
 
-OffPolicyEstimate = DeveloperAPI(
+OffPolicyEstimate = ExperimentalAPI(
     namedtuple("OffPolicyEstimate", ["estimator_name", "metrics"])
 )
 
 
-@DeveloperAPI
+@ExperimentalAPI
```
Comment (sven1977) [inline rllib/offline/estimators/off_policy_estimator.py:19]: same

File: rllib/offline/estimators/off_policy_estimator.py [hunk 10]
```diff
@@ -11,16 +11,16 @@
...
 class OffPolicyEstimator:
     """Interface for an off policy reward estimator."""
 
-    @DeveloperAPI
+    @ExperimentalAPI
```
Comment (sven1977) [inline rllib/offline/estimators/off_policy_estimator.py:23]: same

File: rllib/tests/test_io.py [hunk 11]
```diff
@@ -151,25 +150,25 @@ def test_agent_input_postprocessing_enabled(self):
             self.assertEqual(result["timesteps_total"], 250)  # read from input
             self.assertTrue(np.isnan(result["episode_reward_mean"]))
 
-    def test_agent_input_eval_sim(self):
-        for fw in framework_iterator():
+    def test_agent_input_eval_sampler(self):
+        for fw in ["torch"]:
```
Comment (sven1977) [inline rllib/tests/test_io.py:154]: Please use
```
config = ...

for fw in framework_iterator(config, frameworks="torch"):
```
instead for consistency.

File: rllib/tests/test_io.py [hunk 12]
```diff
@@ -228,17 +226,13 @@ def test_multi_agent(self):
...
             pg.stop()
             pg = PG(
                 env="multi_agent_cartpole",
                 config={
                     "num_workers": 0,
-                    "input": self.test_dir,
-                    "off_policy_estimation_methods": {
-                        "simulation": {"type": "simulation"}
-                    },
+                    "input": self.test_dir + fw,
```
Comment (sven1977) [inline rllib/tests/test_io.py:235]: Quick question: We still support both a) a dir string (read all JSON files in the dir) and b) a list of filenames, correct?

**Review by sven1977 — COMMENTED** [review]
Looks great. Thanks for the cleanup @Rohan138 ! Just a few questions and nits to fix.

File: rllib/evaluation/rollout_worker.py (hunk 3 above)
Comment (Rohan138) [reply rllib/evaluation/rollout_worker.py:762]: I still have the class path as a TODO, will fix that after the Trainer PR, but agree

File: rllib/tests/test_io.py (hunk 11 above)
Comment (Rohan138) [reply rllib/tests/test_io.py]: That was supposed to have TF too, fixed now

File: rllib/evaluation/rollout_worker.py (hunk 4 above)
Comment (avnishn) [reply rllib/evaluation/rollout_worker.py:789]: what is `blackhole_outputs` ?

File: rllib/evaluation/worker_set.py [hunk 13]
```diff
@@ -682,7 +677,7 @@ def valid_module(class_path):
             log_level=config["log_level"],
             callbacks=config["callbacks"],
             input_creator=input_creator,
-            off_policy_estimation_methods=off_policy_estimation_methods,
+            off_policy_estimation_methods=config["off_policy_estimation_methods"],
```
Comment (avnishn) [inline rllib/evaluation/worker_set.py:680]: makes sense to me -- should be able to use OPE even if the input is a sampler (for debug purposes, or if working with a separate eval ds)

File: rllib/offline/estimators/direct_method.py [hunk 14]
```diff
@@ -20,23 +20,27 @@
 logger = logging.getLogger()
 
 
-@ExperimentalAPI
+@DeveloperAPI
 def train_test_split(
```
Comment (avnishn) [inline rllib/offline/estimators/direct_method.py:24]: so where does this get invoked, now that we will require train and test datasets to be specified ahead of time?

**Review by avnishn — COMMENTED** [review]
ok this p much looks good to me. 

I left quite a few questions, if you could answer those first and then lets move forward.

File: rllib/algorithms/algorithm.py [hunk 15]
```diff
@@ -353,6 +353,21 @@ def setup(self, config: PartialAlgorithmConfigDict):
         self.workers: Optional[WorkerSet] = None
         self.train_exec_impl = None
 
+        # Offline RL settings.
+        input_evaluation = config.get("input_evaluation")
+        if input_evaluation is not None and input_evaluation is not DEPRECATED_VALUE:
+            ope_dict = {str(ope): {"type": ope} for ope in input_evaluation}
```
[the remark this answers is not in the document]
Comment (Rohan138) [reply rllib/algorithms/algorithm.py:359]: Agree, I had to use the same format for the q_model as well in the other PR, it's super awkward

File: rllib/evaluation/rollout_worker.py [hunk 16]
```diff
@@ -733,10 +727,14 @@ def wrap(env):
...
-            if name == "simulation":
-                logger.warning(
-                    "Requested 'simulation' input evaluation method: "
-                    "will discard all sampler outputs and keep only metrics."
+            if method_type == "simulation":
+                deprecation_warning(
+                    old='off_policy_estimation_methods={"simulation"}',
+                    new='input="sampler"',
+                    help="The `simulation` estimation method has been deprecated."
+                    "If you want to run online evaluation on your data, use"
```
[the remark this answers is not in the document]
Comment (Rohan138) [reply rllib/evaluation/rollout_worker.py:735]: `simulation` was somehow backed by AsyncSampler with `blackhole_outputs=True`, but yes.

File: rllib/evaluation/worker_set.py (hunk 13 above)
Comment (Rohan138) [reply rllib/evaluation/worker_set.py:680]: Yup, only thing that probably doesn't work currently is mixed input e.g. ("sampler" + json)

File: rllib/offline/estimators/direct_method.py (hunk 14 above)
Comment (Rohan138) [reply rllib/offline/estimators/direct_method.py:24]: The whole function gets deleted in the next PR :) the user should split them ahead of time, or we can make this a utility in Ray Train/Datasets.

**Review by avnishn — APPROVED** [review]

**Comment by avnishn** [comment]
Looks like the failing tests are unrelated. This is ok to merge

**Pull request merged**